ceno_emul/src/host_utils.rs

```diff
@@ -0,0 +1,32 @@
+use std::iter::from_fn;
+
+use crate::{ByteAddr, EmuContext, VMState, WordAddr};
+
+const WORD_SIZE: usize = 4;
+const INFO_OUT_ADDR: WordAddr = ByteAddr(0xC000_0000).waddr();
+
+pub fn read_all_messages(state: &VMState) -> Vec<String> {
+    let mut offset: WordAddr = WordAddr::from(0);
+    from_fn(move || match read_message(state, offset) {
+        out if out.is_empty() => None,
+        out => {
+            offset += out.len().div_ceil(WORD_SIZE) as u32 + 1;
+            Some(out)
+        }
+    })
+    .collect()
+}
+
+fn read_message(state: &VMState, offset: WordAddr) -> String {
+    let out_addr = INFO_OUT_ADDR + offset;
+    let byte_len = state.peek_memory(out_addr) as usize;
+
+    String::from_utf8_lossy(
+        &(out_addr + 1_usize..)
+            .map(|memory| state.peek_memory(memory))
```

Comment:
`.map(|memory| state.peek_memory(memory))`: rename it to `address`?

Reply:
Done.